### PR TITLE
datasource: support apollo

### PR DESCRIPTION
**Pull request by yedamao**
### Describe what this PR does / why we need it
Datasource support apollo

### Special notes for reviews

### CODE REVIEW

**Comment by louyuting** [comment]
@yedamao  The CI indicates failed, could you please resolve it?

**Comment by yedamao** [comment]
@louyuting  done

File: ext/datasource/apollo/apollo.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+func (s *apolloDataSource) Initialize() error {
+	if !s.isInitialized.CompareAndSet(false, true) {
+		return errors.New("Apollo datasource had been initialized")
+	}
+	if err := s.doReadAndUpdate(); err != nil {
+		// Failed to read default should't block initialization
+		logging.Error(err, "[Apollo] Failed to read initial data for namespace", "namespace", s.namespace)
+	}
+
+	go util.RunWithRecover(s.watch)
```
[the remark this answers is not in the document]
Comment (yedamao) [reply ext/datasource/apollo/apollo.go:49]: watch is a go func. How about just log the error

File: ext/datasource/apollo/apollo.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+func (s *apolloDataSource) Close() error {
+	s.client.Stop()
+	close(s.stop)
+	return nil
+}
+
+func (s *apolloDataSource) watch() {
+
+	errChan := s.client.Start()
```
[the remark this answers is not in the document]
Comment (yedamao) [reply ext/datasource/apollo/apollo.go:77]: next step is log it 😂

File: ext/datasource/apollo/apollo.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+func (s *apolloDataSource) ReadSource() ([]byte, error) {
+	raw := s.client.GetNameSpace(s.namespace)
+	v, exist := raw["content"]
+	if !exist {
+		return nil, errors.New("namespace does not exist/empty")
+	}
+
+	str, ok := v.(string)
```
[the remark this answers is not in the document]
Comment (yedamao) [reply ext/datasource/apollo/apollo.go:61]: The namespace type of apollo supported

- properties
-  json
-  yml
-  yaml
-  txt
-  xml

The above namespace types except `properties` v is string type. `properties` not support currently
`json` namespace content works well with json related rules converter in ext/datasource.
`yml/yaml/txt/xml` may need user implement it's own rules converter

File: ext/datasource/apollo/apollo.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sanxun0325) [reply ext/datasource/apollo/apollo.go:77]: If execution of s.client.start() fails, does it make sense to continue executing s.client. WatchNamespace()

File: ext/datasource/apollo/apollo.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sanxun0325) [reply ext/datasource/apollo/apollo.go:49]: If called s.watch func fails in Initialize func, the error message should be returned directly to the caller.

File: ext/datasource/apollo/apollo.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (yedamao) [reply ext/datasource/apollo/apollo.go:77]: The s.client.start() will start a  long poll goroutine and return errChan immediately.

File: ext/datasource/apollo/apollo.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (yedamao) [reply ext/datasource/apollo/apollo.go:49]:
```suggestion
func (s *apolloDataSource) Initialize() error {
        if !s.isInitialized.CompareAndSet(false, true) {
                return errors.New("Apollo datasource had been initialized")
        }
        if err := s.doReadAndUpdate(); err != nil {
                return err
        }

        go util.RunWithRecover(s.watch)

        return nil
}```

how about return err when first call  s.doReadAndUpdate()

[the remark this answers is not in the document]
Comment (sanxun0325) [reply ext/datasource/apollo/apollo.go:49]: It seems to be better this way

File: ext/datasource/apollo/apollo.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	"github.com/alibaba/sentinel-golang/util"
+	"github.com/shima-park/agollo"
+)
+
+type apolloDataSource struct {
+	datasource.Base
+	namespace     string
+	client        agollo.Agollo
+	isInitialized util.AtomicBool
+	stop          chan bool
```
Comment (louyuting) [inline ext/datasource/apollo/apollo.go:18]: Maybe the `chan struct{}` is better。

Comment (yedamao) [reply ext/datasource/apollo/apollo.go:18]: I also think that

**Comment by sczyh30** [comment]
@yedamao Could you please resolve conflicts?

**Comment by yedamao** [comment]
@sczyh30 done.

 I noticed the ext/datasource pkg has been moved to sentinel-go-datasources repo, maybe should I open a pr on there?

**Comment by sczyh30** [comment]
> @sczyh30 done.
> 
> I noticed the ext/datasource pkg has been moved to sentinel-go-datasources repo, maybe should I open a pr on there?

Yes, you may submit a PR in `sentinel-go-datasources` repo.

**Pull request closed without merging**